injector/injector.go

```diff
@@ -0,0 +1,312 @@
+				if isCreative {
+					injectTracker = false
+					encoder.Flush()
+					encoder.EncodeToken(tt)
+					encoder.Flush()
+					for typ, urls := range ti.events.LinearTrackingEvents {
+						ti.provider.PopulateEventMacros(creativeId, "tracking", typ)
+						for _, url := range urls {
+							outputXML.WriteString("<Tracking event=\"")
+							outputXML.WriteString(string(typ))
+							outputXML.WriteString("\"><![CDATA[")
+							ti.replacer.ReplaceBytes(&outputXML, url, ti.provider)
+							outputXML.WriteString("]]></Tracking>")
+						}
+					}
+					continue
+				}
+			}
+
+		case xml.EndElement:
+			switch tt.Name.Local {
+			case "Impression":
+				encoder.Flush()
+				encoder.EncodeToken(tt)
+				encoder.Flush()
+				fmt.Println()
+				if !impressionTagFound {
+					for _, url := range ti.events.Impressions {
+						outputXML.WriteString("<Impression><![CDATA[")
+						ti.replacer.ReplaceBytes(&outputXML, url, ti.provider)
+						outputXML.WriteString("]]></Impression>")
+					}
+					impressionTagFound = true
+				}
+				continue
+			case "Error":
+				encoder.Flush()
+				encoder.EncodeToken(tt)
+				encoder.Flush()
+				if !errorTagFound {
+					for _, url := range ti.events.Errors {
+						outputXML.WriteString("<Error><![CDATA[")
+						ti.replacer.ReplaceBytes(&outputXML, url, ti.provider)
+						outputXML.WriteString("]]></Error>")
+
+					}
+					errorTagFound = true
+				}
+				continue
+			case "NonLinearAds":
+				if injectTracker {
+					injectTracker = false
+					encoder.Flush()
+					outputXML.WriteString("<TrackingEvents>")
+					for typ, urls := range ti.events.LinearTrackingEvents {
+						ti.provider.PopulateEventMacros(creativeId, "", typ)
+						for _, url := range urls {
+
+							outputXML.WriteString("<Tracking event=\"")
+							outputXML.WriteString(typ)
+							outputXML.WriteString("\"><![CDATA[")
+							ti.replacer.ReplaceBytes(&outputXML, url, ti.provider)
+							outputXML.WriteString("]]></Tracking>")
+						}
+					}
+					outputXML.WriteString("</TrackingEvents>")
+
+					if !nonLineaerTagFound && wrapperTagFound {
+						outputXML.WriteString("<NonLinear>")
+						ti.provider.PopulateEventMacros(creativeId, "", "")
+						for _, url := range ti.events.NonLinearClickTracking {
+							outputXML.WriteString("<NonLinearClickTracking><![CDATA[")
+							ti.replacer.ReplaceBytes(&outputXML, url, ti.provider)
+							outputXML.WriteString("]]></NonLinearClickTracking>")
+						}
+						outputXML.WriteString("</NonLinear>")
+					}
+					encoder.EncodeToken(tt)
+				}
+			case "Linear":
+				if injectVideoClicks {
+					injectVideoClicks = false
+					encoder.Flush()
+					outputXML.WriteString("<VideoClicks>")
+					ti.provider.PopulateEventMacros(creativeId, "", "")
+
+					for _, url := range ti.events.VideoClicks {
+
+						outputXML.WriteString("<ClickTracking><![CDATA[")
+						ti.replacer.ReplaceBytes(&outputXML, url, ti.provider)
+						outputXML.WriteString("]]></ClickTracking>")
+					}
+
```

Comment:
Nitpick: Can you remove this blank line?

Reply:
Done